heartbeat/gcp-vpc-move-vip.in

```diff
@@ -0,0 +1,299 @@
+  except KeyError:
+    return ''
+
+
+def get_localhost_alias():
+  net_iface = get_metadata('instance/network-interfaces', {'recursive': True})
+  net_iface = json.loads(net_iface.decode('utf-8'))
+  try:
+    return net_iface[0]['ipAliases'][0]
+  except (KeyError, IndexError):
+    return ''
+
+
+def get_zone(project, instance):
+  request = CONN.instances().aggregatedList(project=project)
+  while request is not None:
+    response = request.execute()
+    zones = response.get('items', {})
+    for zone in zones.values():
+      for inst in zone.get('instances', []):
+        if inst['name'] == instance:
+          return inst['zone'].split("/")[-1]
+    request = CONN.instances().aggregatedList_next(
+        previous_request=request, previous_response=response)
+  raise Exception("Unable to find instance %s" % (instance))
+
+
+def gcp_alias_start(alias):
+  if not alias:
+    sys.exit(OCF_ERR_CONFIGURED)
+  my_alias = get_localhost_alias()
+  my_zone = get_metadata('instance/zone').split('/')[-1]
+  project = get_metadata('project/project-id')
+
+  # If I already have the IP, exit. If it has an alias IP that isn't the VIP,
+  # then remove it
+  if my_alias == alias:
+    logging.info(
+        '%s already has %s attached. No action required' % (THIS_VM, alias))
+    sys.exit(OCF_SUCCESS)
+  elif my_alias:
+    logging.info('Removing %s from %s' % (my_alias, THIS_VM))
+    set_alias(project, my_zone, THIS_VM, '')
+
+  # Loops through all hosts & remove the alias IP from the host that has it
+  hostlist = os.environ.get('OCF_RESKEY_hostlist', '')
+  hostlist.replace(THIS_VM, '')
+  for host in hostlist.split():
+    host_zone = get_zone(project, host)
+    host_alias = get_alias(project, host_zone, host)
+    if alias == host_alias:
+      logging.info(
+          '%s is attached to %s - Removing all alias IP addresses from %s' %
+          (alias, host, host))
+      set_alias(project, host_zone, host, '')
+      break
+
+  # add alias IP to localhost
+  set_alias(
+      project, my_zone, THIS_VM, alias,
+      os.environ.get('OCF_RESKEY_alias_range_name'))
+
+  # Check the IP has been added
+  my_alias = get_localhost_alias()
+  if alias == my_alias:
+    logging.info('Finished adding %s to %s' % (alias, THIS_VM))
+  elif my_alias:
+    logging.error(
+        'Failed to add IP. %s has an IP attached but it isn\'t %s' %
+        (THIS_VM, alias))
+    sys.exit(OCF_ERR_GENERIC)
+  else:
+    logging.error('Failed to add IP address %s to %s' % (alias, THIS_VM))
+    sys.exit(OCF_ERR_GENERIC)
+
+
+def gcp_alias_stop(alias):
+  if not alias:
+    sys.exit(OCF_ERR_CONFIGURED)
+  my_alias = get_localhost_alias()
+  my_zone = get_metadata('instance/zone').split('/')[-1]
+  project = get_metadata('project/project-id')
+
+  if my_alias == alias:
+    logging.info('Removing %s from %s' % (my_alias, THIS_VM))
+    set_alias(project, my_zone, THIS_VM, '')
+
+
+def gcp_alias_status(alias):
+  if not alias:
+    sys.exit(OCF_ERR_CONFIGURED)
```

Comment:
You should move this out of the different functions to a validate-function (we usually run it before the start-action), and also add the action below and in the meta-data.

Reply:
done